Add task for Apify common use cases.

ammeek opened the pull request:
### What changes are being made and why?
I have added four tasks to this plugin:

#### RunActor: 
This task is used to execute an Apify Actor. It allows you to specify the Actor ID, input parameters, and other configurations like memory, timeout, and maximum items. It initiates the Actor run and returns the details of the run.

#### GetLastActorRun: 
This task retrieves details about the most recent run of a specific Apify Actor. It requires the Actor ID and fetches the last run's metadata, such as its status and results.

#### GetDataset: 
This task fetches the dataset associated with a specific Apify Actor run. It uses short polling to retry fetching the dataset if it is not yet available, ensuring the data is retrieved once the Actor has finished processing.

#### SaveDatasetToFile: 
This task saves the dataset from an Apify Actor run to a temporary file. It supports various formats (e.g., JSON, CSV) and retries fetching the dataset if it is empty, ensuring the data is saved once available.

### How the changes have been QAed?

Too verify the behavior of this plugin I ran these example flows:
#### Example flow to run a actor and fetch it result.

```yaml
id: run_actor_and_get_results
namespace: company.team

tasks:
  - id: run_actor
    type: io.kestra.plugin.apify.actor.Run
    actorId: GdWCkxBtKWOsKjdch
    maxItems: 1
    apiToken: API_TOEKN_PLACEHOLDER
    input:
      excludePinnedPosts: false
      hashtags: ["fyp"]
      proxyCountryCode: "None"
      resultsPerPage: 2
      scrapeRelatedVideos: false
      shouldDownloadAvatars: false
      shouldDownloadCovers: false
      shouldDownloadMusicCovers: false
      shouldDownloadSlideshowImages: false
      shouldDownloadSubtitles: false
      shouldDownloadVideos: false
  - id: log_get_last_run_results
    type: io.kestra.plugin.core.log.Log
    message: '{{outputs.run_actor}}'
  - id: get_data_set_raw
    type: io.kestra.plugin.apify.dataset.Save
    datasetId: '{{outputs.run_actor.defaultDatasetId}}'
    apiToken: API_TOEKN_PLACEHOLDER
  - id: log_get_data_set_raw_results
    type: io.kestra.plugin.core.log.Log
    message: '{{outputs.get_data_set_raw}}'
```

Get last actor run and fetch it's results:
``` yaml
id: get_last_actor_run_and_get_dataset
namespace: company.team

tasks:
  - id: run_actor
    type: io.kestra.plugin.apify.dataset.GetLastRun
    actorId: GdWCkxBtKWOsKjdch
    apiToken: API_TOEKN_PLACEHOLDER
  - id: log_get_last_run_results
    type: io.kestra.plugin.core.log.Log
    message: '{{outputs.run_actor}}'
  - id: get_data_set_raw
    type: io.kestra.plugin.apify.dataset.Get
    datasetId: '{{outputs.run_actor.defaultDatasetId}}'
    apiToken: API_TOEKN_PLACEHOLDER
  - id: log_get_data_set_raw_results
    type: io.kestra.plugin.core.log.Log
    message: '{{outputs.get_data_set_raw}}'
```

### Setup Instructions

A free account can be made with [Apify](https://apify.com/) when I created my account I got $5.00 usd of credit for free which I found was more then enough for testing.

- closes https://github.com/kestra-io/kestra/issues/10531

Malaydewangan09 commented on src/main/java/io/kestra/plugin/apify/task/AbstractGetDataset.java:
```diff
@@ -0,0 +1,172 @@
+        title = "Limit",
+        description = "Maximum number of items to return. By default Limit value is set to 1000."
+    )
+    private Property<Integer> limit;
```
For default values, you can do like this
```suggestion
    @Builder.default
    private Property<Integer> limit = Property.ofValue(1000);
```

fdelbrayelle commented:
Hi @ammeek :wave: Great work, thank you!
Do you think it would be possible for you to propose a PR too on Apify Doc so that there would have a doc on this plugin and how to use Apify through Kestra? Something like [this page for n8n](https://github.com/apify/apify-docs/blob/f58cf83825723bd4c994e3fe9f2f3cea09ad3e6c/sources/platform/integrations/workflows-and-notifications/n8n.md?plain=1#L2). Thanks!

fdelbrayelle commented on src/main/java/io/kestra/plugin/apify/task/RunActor.java:
```diff
@@ -0,0 +1,163 @@
+@Plugin(
+    examples = {
+        @Example(
+            title = "Save dataset with id of mecGriFjtDHRNtYOZ as temp file.",
```
Can you remove all ids like "mecGriFjtDHRNtYOZ" from the examples titles please? Because it will end in the final doc for the user. It's better to stay "generic" like "given id", etc.

fdelbrayelle commented on src/main/java/io/kestra/plugin/apify/ApifyConnection.java:
```diff
@@ -0,0 +1,188 @@
+        } catch (IllegalVariableEvaluationException illegalVariableEvaluationException) {
+            logger.error("Error getting API key for Apify: {}", illegalVariableEvaluationException.getMessage());
+            throw illegalVariableEvaluationException;
+        } catch (Exception e) {
```
You may want to do a `completableFuture.completeExceptionally` in the catch

fdelbrayelle commented on src/test/java/io/kestra/plugin/apify/actor/pricinginfo/CommonActorPricingInfoTest.java:
```diff
@@ -0,0 +1,76 @@
+                "startedAt": "2023-10-01T00:00:00Z",
+                "notifiedAboutFutureChangeAt": "2025-07-05T12:05:00.000Z",
+                "notifiedAboutChangeAt": "2025-06-05T12:05:00.000Z",
+                "reasonForChange": "reasonForChange", [IMPLEMENTATION_SPECIFIC_PROPERTIES]
```
Can you add a fake field in this JSON to ensure it doesn't break the whole deserialization?

ammeek commented:
I've addressed the above comments. I am not sure if I should mark the comments as resolved or should the person who left the comments do that after they review the changes again?

Will work on getting the docs mentioned in this comment written in the next couple of days.

> Hi @ammeek 👋 Great work, thank you! Do you think it would be possible for you to propose a PR too on Apify Doc so that there would have a doc on this plugin and how to use Apify through Kestra? Something like [this page for n8n](https://github.com/apify/apify-docs/blob/f58cf83825723bd4c994e3fe9f2f3cea09ad3e6c/sources/platform/integrations/workflows-and-notifications/n8n.md?plain=1#L2). Thanks!

fdelbrayelle commented:
Thank you @ammeek ! We will review your last changes. You can resolve the conversations if you think it was done or just answer if you didn't implement it for some reason to let the discussion continue if needed.

fdelbrayelle commented on src/main/java/io/kestra/plugin/apify/task/GetDataset.java:
```diff
@@ -0,0 +1,97 @@
+        while (true) {
+            List<?> dataset = this.makeCall(runContext, requestBuilder, List.class);
+
+            if (!dataset.isEmpty()) {
```
Could you actually replace this `while (true)` with something we have in Kestra core: `RetryUtils` + `Exponential`. You can see an example [here](https://github.com/kestra-io/kestra/blob/25f5ccc6b5e424dfa950f7add1fc7f05caac8730/script/src/main/java/io/kestra/plugin/scripts/runner/docker/Docker.java#L861).

fdelbrayelle replied:
Yes correct! Thank you!

Malaydewangan09 commented on src/main/java/io/kestra/plugin/apify/task/AbstractGetDataset.java:
```diff
@@ -0,0 +1,171 @@
+        Optional<String> rView = runContext.render(this.view).as(String.class);
+
+        final Map<String, Object> queryParamValues = new HashMap<>(Map.of(
+            "cleanValue", runContext.render(this.clean).as(Boolean.class).orElseThrow(),
```
You can add default values in `orElse()` like this if any.
```suggestion
            "cleanValue", runContext.render(this.clean).as(Boolean.class).orElse(false),
```

ammeek replied:
I've pushed the suggested changes. I was thinking  if we use the `orElse()` here do we still need to use the default builder pattern? If all of the values that have default values already use the default builder pattern then it is not possible for the value in the 'orElse' to be assigned to the local variable. Since all of the input in these tasks are immutable objects could we add a final static variable to define the default value for specific class properties rather than using the default builder pattern?

For example:
``` java
class Join extends Task implements RunnableTask<Join.Output> {
    @NotNull
    Property<String> textOne;
    
    @NotNull
    Property<String> textTwo;
    
    Property<String> delimiter;
    
    private static final String DELIMITER = ",";
    
    @Override
    public Join.Output run(RunContext runContext) throws Exception {
        String rTextOne = runContext.render(textOne).as(String.class).orElseThrow(
            () -> new Exception("textOne is required")
        );
        String rTextTwo = runContext.render(textTwo).as(String.class).orElseThrow(
            () -> new Exception("textTwo is required")
        );
        String rDelimited = runContext.render(delimiter).as(String.class).orElse(DELIMITER);
        
        return new Join.Output(rTextOne + rDelimited + rTextTwo);
    }

    record Output(String text) implements io.kestra.core.models.tasks.Output {
    }
}
```

fdelbrayelle replied:
You can mix the default builder annotation + the `orElse` as a more defensive approach + the const to put at both places yes.

Malaydewangan09 commented on src/main/java/io/kestra/plugin/apify/task/GetDataset.java:
```diff
@@ -0,0 +1,112 @@
+
+                   tasks:
+                     - id: list_runs
+                       type: io.kestra.plugin.apify.task.GetDataset
```
Better to extract this as `apify.dataset.Get`, as we don't generally explicitly add a `task` package.
WDYT?

Malaydewangan09 commented on src/main/java/io/kestra/plugin/apify/task/GetDataset.java:
```diff
@@ -0,0 +1,112 @@
+        )
+    }
+)
+public class GetDataset extends AbstractGetDataset implements RunnableTask<GetDataset.Output> {
```
Better to extract this as `io.kestra.plugin.apify.dataset.Get`, as we don't generally explicitly add a `task` package.
WDYT?

Malaydewangan09 commented on src/main/java/io/kestra/plugin/apify/task/RunActor.java:
```diff
@@ -0,0 +1,161 @@
+        )
+    }
+)
+public class RunActor extends ApifyConnection implements RunnableTask<ActorRun> {
```
Same, better to extract this as `io.kestra.plugin.apify.actor.Run` no?

Malaydewangan09 commented on src/main/java/io/kestra/plugin/apify/ApifyConnection.java:
```diff
@@ -0,0 +1,190 @@
+    private static final String APIFY_API_URL = "https://api.apify.com/v2";
+    private static final String JSON_CONTENT_TYPE = "application/json; charset=UTF-8";
+
+    private Property<String> apiToken;
```
MIssing @NotNull?

ammeek commented:
@Malaydewangan09 and @fdelbrayelle I've pushed changes to address some of your comments above. I've left one of the comment unresolved to see if their are still additional changes needed based on my latest comment.

fdelbrayelle commented on src/main/java/io/kestra/plugin/apify/dataset/GetLastRun.java:
```diff
@@ -0,0 +1,62 @@
+
+                   tasks:
+                     - id: list_runs
+                       type: io.kestra.plugin.apify.actor.GetLastRun
```
Should be `io.kestra.plugin.apify.dataset.GetLastRun`

Can you check all types in all examples please?

fdelbrayelle commented on src/test/java/io/kestra/plugin/apify/task/GetDatasetTest.java:
```diff
@@ -0,0 +1,121 @@
+package io.kestra.plugin.apify.task;
```
You should have the same package than for the implementation and remove `task` here too

fdelbrayelle commented:
Thanks for the doc PR, I'll have a look!
As for this PR I don't see why we shouldn't be able to merge it, so you can keep it open :)

fdelbrayelle commented:
@ammeek Ah you meant [this commit](https://github.com/ammeek/apify-docs/commit/eba13181ee467487195df7065b4477564ccbb920)? Yes you should open a PR on `apify-docs` indeed :)
